akka-actor/src/main/scala/akka/dispatch/affinity/AffinityPool.scala

```diff
@@ -41,112 +54,166 @@ class AffinityPool(numThreads: Int, affinityGroupSize: Int, tf: ThreadFactory) e
 
   private def tryEnqueue(command: Runnable) = {
     // does a mod on the hash code to determine which queue to put task in
-    val queueIndex = Math.abs(command.hashCode()) & (numThreads - 1)
+    val queueIndex = Math.abs(command.hashCode()) & (parallelism - 1)
     workQueues(queueIndex).add(command)
   }
 
-  private def onWorkerExit(w: ThreadPoolWorker): Unit = {
-    locked(mainLock) {
+  /**
+   * Each worker should go through that method while terminating.
+   * In turn each worker is responsible for modifying the pool
+   * state accordingly. For example if this is the last worker
+   * and the queue is empty and we are in a ShuttingDown state
+   * the worker can transition the pool to ShutDown and attempt
+   * termination
+   *
+   * Furthermore, if this worker has experienced abrupt termination
+   * due to an exception being thrown in user code, the worker is
+   * responsible for adding one more worker to compensate for its
+   * own termination
+   *
+   */
+  private def onWorkerExit(w: ThreadPoolWorker, abruptTermination: Boolean): Unit =
+    locked(bookKeepingLock) {
       workers.remove(w)
+      if (workers.isEmpty && !abruptTermination && poolState >= ShuttingDown) {
+        poolState = ShutDown // transition to shutdown and try to transition to termination
+        attemptPoolTermination()
+      }
+      if (abruptTermination && poolState == Running)
+        addWorker(workers, w.q)
     }
+
+  override def execute(command: Runnable): Unit = {
+    if (command == null)
+      throw new NullPointerException
+    if (!(poolState == Running && tryEnqueue(command)))
```

Comment:
@zaharidichev Yeah, I guess we'll have to see if this will be a problem in practice. I guess what we could do is to prolong the draining phase to catch any straggling enqueues.